### ISSUE TITLE
SSL_write() in client not returning on handshake failure: no shared cipher

### ISSUE THREAD

**Issue by mcfreis** [openssl/openssl#2886]
Hi,

I'm not sure how to handle this from within a client setup. I have a client that connects to a server (both using OpenSSL 1.0.2j). The client wants a curve P-384 where as the server only allwos P-256 (for example).
The server can handle this handshake failure and returns "error:1408A0C1:SSL routines:ssl3_get_client_hello:no shared cipher" and also sends an alert to the client. The client is stuck in ssl_write() and never returns. Shouldn't it return and also notice the "no shared cipher"?

What can I do?

Kind regards
Björn

**Comment by mattcaswell** [comment]
Yes, it should return. Doesn't the server close the underlying connection, and if so is the client still stuck after that? Does the client use blocking IO? Also, is the client using SSL_MODE_AUTO_RETRY? Are you able to attach a debugger and work out where in OpenSSL the client is hanging?

**Comment by mcfreis** [comment]
I forgot to mention, I'm using the DTLS methods. On the server side, I just have the main socket at this point of time and I can't close this one. Maybe I'm getting something wrong there. The client is blocking (same behaviour with the openssl s_client). About SSL_MODE_AUTO_RETRY, I'm not sure, I did not touch this setting. I can't debug into the library.

Any suggestions?

Regards
Björn

**Comment by mattcaswell** [comment]
> On the server side, I just have the main socket at this point of time and I can't close this one. Maybe I'm getting something wrong there.

I'm not quite sure what you mean by that. Why can't you close it?

Does the hang happen consistently every time, or is it occasionally? If occasionally then perhaps the alert is getting lost

How do you know that the server sends an alert? Can you do a wireshark packet capture?

**Comment by mcfreis** [comment]
What I meant, is that the socket is the one, where the client signals first contact and then an accept is done by the server and the the socket gets copied with the corresponding ssl context. If I close this socket, I'll close the server down.

This happens consistently. I did a wireshark capture and attached it. In the wireshark capture, the alert is clearly seen.

![no_cipher_hangs_client](https://cloud.githubusercontent.com/assets/24935824/23802837/dcf620a0-05b4-11e7-948e-64006eaa367d.png)

**Comment by mattcaswell** [comment]
Are you able to supply the original pcap file rather than just an image?

**Comment by mattcaswell** [comment]
BTW are you using DTLSv1_listen()?

**Comment by mattcaswell** [comment]
Also was the capture done on the server side or the client side (or somewhere in the middle)? It would be good to confirm that the alert is actually arriving at its destination and not getting lost somewhere en route.

**Comment by mcfreis** [comment]
Here is the original capture done with wireshark on the server side.

[no_cipher_hangs_client.zip](https://github.com/openssl/openssl/files/834818/no_cipher_hangs_client.zip)

Dropping a few messages would be okay, but all of them?! And yes I'm using DTLSv1_listen() on the server side.

**Comment by mattcaswell** [comment]
I have managed to recreate this issue and I can confirm that this is due to a bug in DTLSv1_listen().

This only impacts 1.0.2. Version 1.1.0 works properly due to the fact that DTLSv1_listen() was completely rewritten for that version.

I have a proposed fix in #2915.

DTLSv1_listen() is supposed to operate statelessly. That means it handles record sequence numbers differently. Specifically the write sequence number for the record is supposed to reflect the sequence number we read in from the client. The logic for doing this occurs too late, after we have finished processing the ClientHello. Therefore if an alert gets sent before that then it is sent with whatever sequence number happens to be set.

My guess is that we have not seen this problem before because, it turns out, that most of the time the sequence number happens to be correct by chance. However there are a number of scenarios which could cause you to hit this bug.

Normally when you call DTLSv1_listen() you do so in a simple loop which looks something like this (error handling code omitted for clarity):

````
s = SSL_new();
do {
    ret = DTLSv1_listen(s);
} while (ret <= 0);
````

One scenario which could cause you to reliably hit this bug is to do it like this instead:
````
do {
    s = SSL_new();
    ret = DTLSv1_listen(s);
    SSL_free(s);
} while (ret <= 0);
````
The creation of a new SSL object ever time you go around the loop is slightly unusual, and not strictly required, but it should work. However it has the the side effect of resetting the sequence numbers, meaning that you reliably hit this bug. There are probably other ways to hit the bug too.
